### PR TITLE
For #10173: login duplicates and save

### PR DESCRIPTION
**Pull request by eliserichards**
For #10173
Related to https://github.com/mozilla/application-services/issues/3330 - fix in https://github.com/mozilla/application-services/pull/3331
Requires https://github.com/mozilla-mobile/android-components/pull/7694

### Requirements for this PR
- Save button enabled/disabled based on errors
- Pull list of duplicates and don't allow dupes to be saved
- UX cleanup for login detail and edit screens
- Move AC `components.core.passwordsStorage` calls into their own store: `LoginsDataStore`
  - to be tested

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt [hunk 1]
```diff
@@ -4,90 +4,129 @@
...
 import org.mozilla.fenix.ext.settings
+import org.mozilla.fenix.theme.ThemeManager
+
+/**
+ * Interactor for the edit login screen.
+ */
+class EditLoginInteractor(private val editLoginController: EditSavedLoginsController) {
+    fun findDuplicates(item: SavedLogin) {
+        // What scope should be used here?
+        GlobalScope.launch(IO) { editLoginController.findPotentialDuplicates(item) }
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt:48]: you can launch on IO off the viewLifecycleScope

Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt:48]: should the viewLifecycleScope be passed here from the view?

Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt:48]: (nevermind, found an example I can use)

**Comment by ekager** [comment]
Can you comment the filed out data review form here so we can get that review rolling too?

**Comment by eliserichards** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) **What questions will you answer with this data?**
- What % of our users use CRUD actions (create, read, update, delete) on their saved logins?
- How frequently do users manipulate their saved logins?

2) **Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?**

- It is crucial we know what types of user segments we have so we can prioritize future logins features. This data will help us better slice our user population & figure out retention metrics. It will also help us to understand the relationship between the Lockwise userbase (standalone password manager) and the Fenix browser users who use saved logins.

3) **What alternative methods did you consider to answer these questions? Why were they not sufficient?**

* There are currently no alternatives.

4) **Can current instrumentation answer these questions?**

* No, there are no currently metrics around these actions.

5) **List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data c](https://wiki.mozilla.org/Firefox/Data_Collection)[ollection ](https://wiki.mozilla.org/Firefox/Data_Collection)[categories](https://wiki.mozilla.org/Firefox/Data_Collection) on the found on the Mozilla wiki.**

* All data is Category 2.

6) **How long will this data be collected?**

* Until 09/01/2020

7) **What populations will you measure?**

* All release, beta, and nightly users with telemetry enabled.

8) **Please provide a general description of how you will analyze this data.**

* Glean / Amplitude

9) **Where do you intend to share the results of your analysis?**

* Only on Glean, Amplitude and with mobile teams.

File: app/metrics.yaml [hunk 2]
```diff
@@ -1988,6 +1988,39 @@ logins:
...
       - fenix-core@mozilla.com
     expires: "2020-09-01"
+  edit_saved_login:
+    type: event
+    description: |
+      A user viewed the edit screen for an individual saved login
+    bugs:
+      - https://github.com/mozilla-mobile/fenix/issues/10173
+    data_reviews:
+      - https://github.com/mozilla-mobile/fenix/issues/10173
```
Comment (ekager) [inline app/metrics.yaml:1998]: Make sure to update these data review fields to link to this PR!

File: app/metrics.yaml [hunk 3]
```diff
@@ -1988,6 +1988,39 @@ logins:
     notification_emails:
       - fenix-core@mozilla.com
     expires: "2020-09-01"
+  edit_saved_login:
+    type: event
+    description: |
+      A user viewed the edit screen for an individual saved login
```
Comment (liuche) [inline app/metrics.yaml:1994]: Might be more specfic, "A user entered the edit screen..."

File: app/metrics.yaml [hunk 4]
```diff
@@ -1988,6 +1988,39 @@ logins:
     notification_emails:
       - fenix-core@mozilla.com
     expires: "2020-09-01"
+  edit_saved_login:
```
Comment (liuche) [inline app/metrics.yaml:1991]: I'd suggest renaming this, because the "edit" could be construed as an action when seen by DS in isolation, like `open_login_editor` or something like that.

File: app/src/main/res/color/save_enabled_ic_color.xml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/res/color/save_enabled_ic_color.xml:1]: This has different resource values than the `primary_state_list_text_color`

**Comment by eliserichards** [comment]
Pulled out the string into #11516 and tagged you for approval @ekager 🔥

File: app/src/main/java/org/mozilla/fenix/settings/logins/LoginsDataStore.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import mozilla.components.concept.storage.Login
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.settings.logins.fragment.EditLoginFragmentDirections
+
+open class LoginsDataStore(
```
Comment (eliserichards) [inline app/src/main/java/org/mozilla/fenix/settings/logins/LoginsDataStore.kt:21]: Move this to a controller

File: app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt [hunk 7]
```diff
@@ -1,307 +0,0 @@
...
-import org.mozilla.fenix.components.metrics.Event
-import org.mozilla.fenix.ext.components
-import org.mozilla.fenix.ext.redirectToReAuth
-import org.mozilla.fenix.ext.settings
-
-/**
- * Displays the editable saved login information for a single website.
- */
-@Suppress("TooManyFunctions", "NestedBlockDepth", "ForbiddenComment")
-class EditLoginFragment : Fragment(R.layout.fragment_edit_login) {
```
Comment (eliserichards) [inline app/src/main/java/org/mozilla/fenix/settings/logins/EditLoginFragment.kt:45]: This was not deleted, just moved to /logins/fragment package

**Comment by eliserichards** [comment]
Waiting on https://github.com/mozilla/application-services/pull/3331 to land, then this should be g2g

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    private val viewLifecycleScope: CoroutineScope,
+    private val navController: NavController,
+    private val loginsFragmentStore: LoginsFragmentStore
+) {
+    private val activity: HomeActivity = context as HomeActivity
+
+    private suspend fun getLogin(loginId: String): Login? =
+        activity.components.core.passwordsStorage.get(loginId)
+
+    fun delete(loginId: String) {
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:43]: Could we give these slightly more descriptive names for easier parsing elsewhere?

File: app/src/main/java/org/mozilla/fenix/settings/logins/interactor/SavedLoginsInteractor.kt [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+ */
+class SavedLoginsInteractor(
+    private val loginsListController: LoginsListController,
+    private val savedLoginsStorageController: SavedLoginsStorageController,
+    private val itemClicked: (SavedLogin) -> Unit,
+    private val learnMore: () -> Unit
+) {
+    fun itemClicked(item: SavedLogin) {
+        itemClicked.invoke(item)
+    }
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/interactor/SavedLoginsInteractor.kt:23]: nit - can we add some space between these funs to make it easier to read?

File: app/src/main/java/org/mozilla/fenix/settings/logins/view/EditLoginView.kt [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * View that contains and configures the Edit Login screen
+ */
+@Suppress("ForbiddenComment")
+class EditLoginView(
+    override val containerView: ViewGroup
+) : LayoutContainer {
+    private val context = containerView.context
+
+    // TODO: create helper class for toggling passwords. Used in login info and edit fragments.
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/view/EditLoginView.kt:25]: Let's file an issue for this if we haven't already and link to it in the comment

File: app/src/main/java/org/mozilla/fenix/settings/logins/view/EditLoginView.kt [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        // For the new type to take effect you need to reset the text to it's current edited version
+        containerView.passwordText?.text = currText
+    }
+
+    fun hidePassword() {
+        val currText = containerView.passwordText?.text
+        containerView.passwordText?.inputType =
+            InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
+        containerView.revealPasswordButton?.setImageDrawable(
+            ResourcesCompat.getDrawable(context.resources,
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/view/EditLoginView.kt:49]: elsewhere we use `AppCompatResources.getDrawable`

File: app/src/main/java/org/mozilla/fenix/settings/logins/view/EditLoginView.kt [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+import kotlinx.android.extensions.LayoutContainer
+import kotlinx.android.synthetic.main.fragment_edit_login.view.*
+import org.mozilla.fenix.R
+import org.mozilla.fenix.components.metrics.Event
+import org.mozilla.fenix.ext.components
+
+/**
+ * View that contains and configures the Edit Login screen
+ */
+@Suppress("ForbiddenComment")
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/view/EditLoginView.kt:19]: Is this because of the added TODO?

File: app/src/main/java/org/mozilla/fenix/settings/logins/view/LoginDetailView.kt [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        webAddressText.text = login.currentItem?.origin
+        usernameText.text = login.currentItem?.username
+        passwordText.text = login.currentItem?.password
+    }
+
+    fun togglePasswordReveal(show: Boolean) {
+        if (show) showPassword() else { hidePassword() }
+    }
+
+    // TODO: create helper class for toggling passwords. Used in login detail and edit fragments.
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/settings/logins/view/LoginDetailView.kt:34]: same nit for TODO here add an issue pls!

File: app/src/main/java/org/mozilla/fenix/settings/logins/view/LoginDetailView.kt [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    fun showPassword() {
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/settings/logins/view/LoginDetailView.kt:35]: I think this should be private.
This and the below should definitely not be publicly exposed to let others modify the View from outside.

But more than this, this shouldn't really exist in a MVI View.
The State received in the `update` method should be structured in such a way to allow the View to map it directly to it's properties. There is a 1-1 relation between them, it's ok for the State to be as granular and specialized as needed.
In this case, if we want such granularity for the passwordText the state should contain something like
- passwordTextToggleImage
- passwordTextContentDescription
- PasswordTextActualText

All of these being propagated through an Action from the Controller which should be the only one knowing about this logic - what should be displayed, when. 
Views should not be left to decide / choose what to display. They should render the received state.

File: app/src/main/java/org/mozilla/fenix/settings/logins/view/LoginDetailView.kt [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+                )
+            )
+            revealPasswordButton.contentDescription =
+                context.resources.getString(R.string.saved_login_hide_password)
+        }
+        // For the new type to take effect you need to reset the text
+        passwordText.text = containerView.passwordText.editableText
+    }
+
+    fun hidePassword() {
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/fenix/settings/logins/view/LoginDetailView.kt:52]: Same as the above.

**Comment by Mugurell** [comment]
One other important improvement to be made:
Since there are 38 files here it is pretty hard to properly review this, there is a high risk un bugs, and afterwards it will be pretty hard to follow why something changed.
I see 21 commits, there will possibly be more followed by a merge of them all at the end, with.. 38 files.
Would be great to have the code split into smaller commits, each addressing a specific need, each having a clear commit message detailing why the change. And then for every fixup needed, like something being raised in review do a rebase-fixup.
I for one very much trust in this guidelines - https://github.com/google/eng-practices/blob/master/review/developer/index.md

**Comment by eliserichards** [comment]
@Mugurell I agree with your analysis. However, this has been pushed off from merge for months now and is not able to be broken down. Regarding the number of comments, it is not ready for review yet so I cannot control the amount of reviews people leave on the PR. I would like this merged and will address bugs as needed. CC: @liuche

File: app/src/main/java/org/mozilla/fenix/settings/logins/view/EditLoginView.kt (hunk 11 above)
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/view/EditLoginView.kt:49]: This was a lint or detekt error telling me to use ResourcesCompat. I can use AppCompatResources instead :)

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+/**
+ * Controller for all saved logins interactions with the password storage component
+ */
+open class SavedLoginsStorageController(
+    context: Context,
+    private val viewLifecycleScope: CoroutineScope,
+    private val navController: NavController,
+    private val loginsFragmentStore: LoginsFragmentStore
+) {
+    private val activity: HomeActivity = context as HomeActivity
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:38]: This is a flow taken from a number of other places in the app. I would not like to change it without cleaning up every other place as well. Maybe we can file an eng health ticket.

File: app/src/main/java/org/mozilla/fenix/settings/logins/fragment/SavedLoginsFragment.kt [hunk 17]
```diff
@@ -83,16 +87,37 @@ class SavedLoginsFragment : Fragment() {
...
+            )
+
         savedLoginsInteractor =
-            SavedLoginsInteractor(savedLoginsController, ::itemClicked, ::openLearnMore)
-        savedLoginsView = SavedLoginsView(view.savedLoginsLayout, savedLoginsInteractor)
-        loadAndMapLogins()
+            SavedLoginsInteractor(
+                loginsListController,
+                savedLoginsStorageController,
+                ::itemClicked,
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/fragment/SavedLoginsFragment.kt:113]: Not within scope of edit logins dupe checking

File: app/src/main/java/org/mozilla/fenix/settings/logins/fragment/LoginDetailFragment.kt [hunk 18]
```diff
@@ -71,29 +72,44 @@ class LoginDetailFragment : Fragment(R.layout.fragment_login_detail) {
...
 
         return view
     }
 
     @ObsoleteCoroutinesApi
     @ExperimentalCoroutinesApi
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        interactor = LoginDetailInteractor(
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/fragment/LoginDetailFragment.kt:92]: This was intentional

File: app/src/main/java/org/mozilla/fenix/settings/logins/LoginsFragmentStore.kt [hunk 19]
```diff
@@ -54,11 +54,11 @@ sealed class LoginsAction : Action {
     data class UpdateLoginsList(val list: List<SavedLogin>) : LoginsAction()
     data class UpdateCurrentLogin(val item: SavedLogin) : LoginsAction()
     data class SortLogins(val sortingStrategy: SortingStrategy) : LoginsAction()
+    data class ListOfDupes(val dupeList: List<SavedLogin>) : LoginsAction()
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/LoginsFragmentStore.kt:57]: I prefer this name. If possible duplicates exist, this list is non-empty. The action does not depend on the list changing; it will be changing with every edit.

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:38]: https://github.com/mozilla-mobile/fenix/blob/f42465657575f94e1d02e0ab95ee56581d3d1676/app/src/main/java/org/mozilla/fenix/settings/advanced/DefaultLocaleSettingsController.kt#L19

https://github.com/mozilla-mobile/fenix/blob/f42465657575f94e1d02e0ab95ee56581d3d1676/app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkController.kt#L52

https://github.com/mozilla-mobile/fenix/blob/f42465657575f94e1d02e0ab95ee56581d3d1676/app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarController.kt#L63

https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/tabtray/TabTrayController.kt#L34

File: app/src/main/java/org/mozilla/fenix/library/history/HistoryFragment.kt [hunk 20]
```diff
@@ -210,9 +210,10 @@ class HistoryFragment : LibraryPageFragment<HistoryItem>(), UserInteractionHandl
         return if (historyItems.size > 1) {
             getString(R.string.history_delete_multiple_items_snackbar)
         } else {
-            getString(
-                R.string.history_delete_single_item_snackbar,
-                historyItems.first().url.toShortUrl(requireComponents.publicSuffixList)
+            String.format(
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/library/history/HistoryFragment.kt:213]: lint/detekt error

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (ekager) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:38]: It is a common pattern that we can file a followup ticket, but in this case we really are only using the context to get the `passwordsStorage`, so instead of passing in the context, we can just pass in the `passwordsStorage` directly.

File: app/src/main/java/org/mozilla/fenix/settings/logins/interactor/SavedLoginsInteractor.kt [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Interactor for the saved logins screen
+ */
+class SavedLoginsInteractor(
+    private val loginsListController: LoginsListController,
+    private val savedLoginsStorageController: SavedLoginsStorageController,
+    private val itemClicked: (SavedLogin) -> Unit,
+    private val learnMore: () -> Unit
+) {
+    fun itemClicked(item: SavedLogin) {
```
[the remark this answers is not in the document]
Comment (ekager) [reply app/src/main/java/org/mozilla/fenix/settings/logins/interactor/SavedLoginsInteractor.kt:21]: I agree with these naming nits. We're already modifying these files, let's just get the naming in line with the rest of the codebase.

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:38]: Changed this to use context instead of HomeActivity, and filed a followup: https://github.com/mozilla-mobile/fenix/issues/12565

[the remark this answers is not in the document]
Comment (ekager) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt:38]: I think we can still improve this further by changing it pass `passwordsStorage` directly since AFAICT that's all we use it for

[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt]: is it cool to pass a component around like that?

File: app/src/main/java/org/mozilla/fenix/settings/logins/fragment/LoginDetailFragment.kt [hunk 22]
```diff
@@ -71,29 +72,44 @@ class LoginDetailFragment : Fragment(R.layout.fragment_login_detail) {
...
+            SavedLoginsStorageController(
+                context = requireContext(),
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/settings/logins/fragment/LoginDetailFragment.kt:97]: Wondering if this might warrant a larger cleanup in the codebase...

File: app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (ekager) [reply app/src/main/java/org/mozilla/fenix/settings/logins/controller/SavedLoginsStorageController.kt]: Ya! See https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/collections/CollectionCreationController.kt#L76

File: app/metrics.yaml [hunk 23]
```diff
@@ -2444,6 +2444,39 @@ logins:
...
       - fenix-core@mozilla.com
     expires: "2020-10-01"
+  open_login_editor:
+    type: event
+    description: |
+      A user entered the edit screen for an individual saved login
+    bugs:
+      - https://github.com/mozilla-mobile/fenix/issues/10173
+    data_reviews:
+      - https://github.com/mozilla-mobile/fenix/issues/11208
```
Comment (ekager) [inline app/metrics.yaml:2454]: nit typo - these should link to the PR with /fenix/pull/11208

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 24]
```diff
@@ -26,6 +26,7 @@ import androidx.navigation.fragment.NavHostFragment
 import androidx.navigation.ui.AppBarConfiguration
 import androidx.navigation.ui.NavigationUI
 import androidx.recyclerview.widget.LinearLayoutManager
+import com.google.android.gms.tasks.Tasks.call
```
Comment (ekager) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:29]: extraneous import?

**Comment by ekager** [comment]
test-debug failing on `WHEN finding login dupes, THEN update duplicates in the store`
